CBinding/CLang/CLangManager.cs

```diff
@@ -0,0 +1,528 @@
+				proj.db.Reset (fileName);
+				CXCursor TUcursor = clang.getTranslationUnitCursor (TU);
+				TranslationUnitParser parser = new TranslationUnitParser (proj.db, fileName, cancellationToken, TUcursor);
+				clang.visitChildren (TUcursor, parser.Visit, new CXClientData (new IntPtr (0)));
+			}
+		}
+
+		/// <summary>
+		/// Removes a translation unit from the collection and disposes its unmanaged resources.
+		/// </summary>
+		/// <param name="proj">
+		/// A <see cref="CProject"/> reference: the project which the TU is associated.
+		/// </param>
+		/// <param name="fileName">
+		/// A <see cref="string"/>: The filename associated with the TU. Basically the source file's name
+		/// </param>
+		public void RemoveTranslationUnit (CProject proj, string fileName)
+		{
+			lock (SyncRoot) {
+				clang.disposeTranslationUnit (translationUnits [fileName]);
+				translationUnits.Remove (fileName);
+			}
+		}
+
+		/// <summary>
+		/// Update Translation units with the correct compiler arguments. Subscribe to event: Project.DefaultConfigurationChanged
+		/// </summary>
+		public void CompilerArgumentsUpdate (object sender, EventArgs args) 
+		{
+			lock (SyncRoot) {
+				if (project.Loading)
+					//on project load its unnecessary to update this, because creating the TU's are already
+					//done with the first active configuration - also doing so sometimes results in an exception
+					return;
+				ClangCCompiler compiler = new ClangCCompiler ();
+				CProjectConfiguration active_configuration =
+					(CProjectConfiguration)project.GetConfiguration (IdeApp.Workspace.ActiveConfiguration);
+				while (active_configuration == null) {
+					Thread.Sleep (20);
+					active_configuration = (CProjectConfiguration)project.GetConfiguration (IdeApp.Workspace.ActiveConfiguration);
+				}
+				string[] compilerArgs = compiler.GetCompilerFlagsAsArray (project, active_configuration);
+				foreach (var TU in translationUnits) {
+					clang.disposeTranslationUnit (translationUnits [TU.Key]);
+					translationUnits [TU.Key] = clang.parseTranslationUnit (
+						index,
+						TU.Key,
+						compilerArgs,
+						compilerArgs.Length,
+						//CDocumentParser.Parse will reparse with unsaved files, risky to get them from here
+						null,
+						0,
+						clang.defaultEditingTranslationUnitOptions ()
+					);
+				}
+			}
+		}
+
+		/// <summary>
+		/// Code completion wrapper to expose clang_codeCompleteAt and handle threading locks-issues.
+		/// </summary>
+		/// <param name="documentContext">
+		/// A <see cref="DocumentContext"/> reference: the document context of the code completion request.
+		/// </param>
+		/// <param name="unsavedFiles">
+		/// A <see cref="CXUnsavedFile"/> array: The unsaved files in the IDE. Obligatory to have valid suggestions.
+		/// </param>
+		/// <param name = "editor">
+		/// A <see cref="CTextEditorExtension"/>: the editor, which edits the document. Contains caret positions, etc.
+		/// </param>
+		public IntPtr codeComplete (
```

Comment:
It's not obvious what the IntPtr's ownership and thread safety is here. Would be good to address in a comment, at least.